x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -664,7 +664,8 @@ private void authorizeRunAs(
         final ActionListener<AuthorizationResult> listener
     ) {
         final Authentication authentication = requestInfo.getAuthentication();
-        if (authentication.getLookedUpBy() == null) {
+        assert authentication.isRunAs() : "authentication does not have a run-as";
+        if (authentication.getEffectiveSubject().getRealm() == null) {
```

Comment:
This is a slight change in business logic: if `runAs` is `false`, the condition here would be `true` regardless of the value of `authentication.getEffectiveSubject().getRealm()` before the change, whereas now this is not the case anymore (in production, where assertions are disabled). Just double-checking that this is intentional.

Reply:
The enclosing method is `authorizeRunAs` which is called inside `maybeAuthorizeRunAs` [when the authentication is indeed a run-as](https://github.com/elastic/elasticsearch/blob/c19f642d94a7bce62303c604f55d6b7783df39f2/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java#L350-L379). So the new assertion is basically making the existing invariant explicit and does not change existing logic.

That said, I will change the assertion message to something like 
> authentication must have run-as for run-as to be authorized

Hope it makes the intention clearer.

Alternatively, we can inline this method. But since it is private. I don't feel strong for it. What do you think?

Reply:
I missed the fact that `isRunAs` is necessarily true here due to `maybeAuthorizeRunAs`. I had only looked at it in isolation. I'm good to leave as is -- it's a private method and to anyone looking at the code as a whole (as opposed to just the diff in the PR), things will make sense. I don't think we should inline.